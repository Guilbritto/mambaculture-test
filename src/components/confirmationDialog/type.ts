
export type ConfirmationDialogProps = {
    callback: () => void
    label: string

}