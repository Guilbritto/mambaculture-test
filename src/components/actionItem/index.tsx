import { ActionItemProps } from "@/components/actionItem/types"

export const ActionItem = ({label,icon, onClick}: ActionItemProps) => {
    
    return (
        <div onClick={onClick} 
            className="border-b flex items-center gap-2 p-3 cursor-pointer hover:bg-slate-100 font-normal space-x-2">
            {icon}{label}
        </div>
    )
}