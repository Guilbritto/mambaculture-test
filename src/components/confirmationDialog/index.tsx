import { ConfirmationDialogProps } from "@/components/confirmationDialog/type"
import { Button } from "@/components/ui/button"
import { useModal } from "@/context/useModal"

export const ConfirmationDialog = ({callback, label}: ConfirmationDialogProps) => {
    const { closeModal} = useModal()

    return (
        <div>
            <h1 className="text-lg font-semibold pb-3">Atenção!</h1>
            <p className="text-muted-foreground text-sm font-normal">{label}</p>
            <div className="flex justify-end gap-2 mt-5">
                <Button onClick={() => closeModal()} variant="outline">Cancelar</Button>
                <Button onClick={() => callback()}>Continuar</Button>
            </div>
        </div>
    )
}
