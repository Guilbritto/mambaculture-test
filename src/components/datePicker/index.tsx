"use client"

import * as React from "react"
import { CalendarIcon } from "@radix-ui/react-icons"
import { format  } from "date-fns"
import { ptBR } from "date-fns/locale"
import { cn } from "@/lib/utils"
import { Button } from "@/components/ui/button"
import { Calendar } from "@/components/ui/calendar"
import {
  Popover,
  PopoverContent,
  PopoverTrigger,
} from "@/components/ui/popover"

type DatePickerProps = {
    label: string
    selected: Date | null
    onSelect: (date?: Date) => void
}

export function DatePicker({label, onSelect, selected}:DatePickerProps) {
  return (
    <Popover>
      <PopoverTrigger asChild>
        <Button
          variant={"outline"}
          className={cn(
            "w-[220px] justify-start text-left font-normal",
            !selected && "text-muted-foreground"
          )}
        >
          <CalendarIcon className="mr-2 h-4 w-4" />
          {selected ? (
            format(selected, 'PPP', { locale: ptBR })
          ) : (
            <span>{label || 'Selecione uma data'} </span>
          )}
        </Button>
      </PopoverTrigger>
      <PopoverContent className="w-auto p-0" align="start">
        <Calendar
          disabled={{ before: new Date() }} 
          mode="single"
          className="bg-transparent"
          selected={selected || undefined}
          onSelect={onSelect}
          initialFocus
        />
      </PopoverContent>
    </Popover>
  )
}
