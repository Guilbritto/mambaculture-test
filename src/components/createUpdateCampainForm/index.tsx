'use client'
import { Form, FormControl, FormField, FormItem, FormLabel, FormMessage } from '@/components/ui/form'
import { useForm } from 'react-hook-form'
import { z } from "zod"
import { zodResolver } from "@hookform/resolvers/zod"
import { Input } from '@/components/ui/input'
import { Button } from '@/components/ui/button'
import { DatePicker } from '@/components/datePicker'
import { format, formatISO } from 'date-fns'
import { Campain } from '@/types/campain'
import { useModal } from '@/context/useModal'
import { useToast } from '@/hooks/use-toast'
import { createCampain, editCampain } from '@/app/actions'
import { useCampain } from '@/context/useCampain'


export type CreateUpdateCampainFormProps = {
    campain?: Campain
}
export const CreateUpdateCampainForm = ({campain}: CreateUpdateCampainFormProps) => {
    const {closeModal} = useModal()
    const { toast } = useToast()
    const {fetchCampains} = useCampain()
    const formSchema: z.ZodSchema = z.object({
        name: z
            .string()
            .min(2, 'O nome deve conter pelo menos 2 caracteres')
            .max(50, 'O nome deve conter no máximo 50 caracteres'),
        startDate: z.date({message: 'Obrigatório informar uma data Inicial'}),
        endDate: z.date({ message: 'Obrigatório informar uma data Final'}).refine((endDate) => {
            return endDate > form.getValues().startDate;
        }, {
            message: 'A data final deve ser maior que a data inicial'
        })
    })

    const form = useForm<z.infer<typeof formSchema>>({
        resolver: zodResolver(formSchema),
        defaultValues: {
            name: campain?.name ?? '',
            startDate: campain ?  new Date(campain?.startDate ?? '') : null,
            endDate: campain ?  new Date(campain?.endDate ?? '') : null ,
        },
    })

    async function onSubmit(values: z.infer<typeof formSchema>) {
        const newCampain: Omit<Campain, 'id'> = {
            name: values.name,
            startDate: values.startDate,
            endDate: values.endDate,
            status: 'active',
            createAt: formatISO(new Date()),
        }

        let response;

        if(campain){
            response = await editCampain({id: campain.id, ...newCampain})
        }else{
            response = await createCampain(newCampain)
        }

        if (!response) {
            toast({
                title: "Erro",
                variant: "destructive",
                description: "Não foi possível salvar as informações, tente novamente mais tarde",
            })
            return
        }

        toast({
            title: "Sucesso",
            description: "Dados salvos com sucesso",
        })
        fetchCampains()
        closeModal()
    }

    return (
        <Form {...form}>
            <h1 className='text-xl text-black pb-3'>
                {campain ? 'Editar campanha' : 'Crie uma nova campanha'}
            </h1>
            <form className='space-y-2' action="" onSubmit={form.handleSubmit(onSubmit)}>
                <FormField
                    control={form.control}
                    name="name"
                    render={({field}) => (
                    <FormItem>

                        <FormLabel>Nome da campanha</FormLabel>
                        <FormControl>
                            <Input placeholder='Nome' {...field} />
                        </FormControl>
                        <FormMessage />
                    </FormItem>
                    )}
                />
                <div className='flex gap-5'>
                    <FormField
                        control={form.control}
                        name="startDate"
                        render={({ field }) => {
                        return (
                            <FormItem>
                                <FormLabel>Data Inicial</FormLabel>
                                <FormControl>
                                    <DatePicker
                                    label={
                                        field.value
                                        ? format(field.value, 'PPP')
                                        : 'Data inicial'
                                    }
                                    selected={field.value}
                                    onSelect={field.onChange}
                                    />
                                </FormControl>
                                <FormMessage />
                            </FormItem>
                        )
                        }}
                    />
                    <FormField
                        control={form.control}
                        name="endDate"
                        render={({ field }) => {
                        return (
                            <FormItem>
                                <FormLabel>Data Final</FormLabel>
                                <FormControl>
                                    <DatePicker
                                        label={
                                            field.value
                                            ? format(field.value, 'PPP')
                                            : 'Data final'
                                        }
                                        selected={field.value}
                                        onSelect={field.onChange}

                                    />
                                </FormControl>
                                <FormMessage />
                            </FormItem>
                        )
                        }}
                    />
                </div>
                <div className='w-full flex justify-end pt-3'>
                    <Button type="submit" className='w-20 bg-slate-800'>
                        Salvar
                    </Button>
                </div>
            </form>
        </Form>
    )
}

