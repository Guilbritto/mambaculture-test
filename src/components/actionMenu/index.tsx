import { removeCampain } from "@/app/actions"
import { ActionItem } from "@/components/actionItem"
import { ActionMenuProps } from "@/components/actionMenu/type"
import { ConfirmationDialog } from "@/components/confirmationDialog"
import { CreateUpdateCampainForm } from "@/components/createUpdateCampainForm"
import { Popover, PopoverContent, PopoverTrigger } from "@/components/ui/popover"
import { useCampain } from "@/context/useCampain"
import { useModal } from "@/context/useModal"
import { EllipsisIcon, Edit, Trash } from "lucide-react"

export const ActionMenu = ({id}: ActionMenuProps) => {
    const {render, openModal, closeModal} = useModal()
    const {fetchCampains} = useCampain()
    const {campains} = useCampain()

    const handleDelete = () => {
        render(<ConfirmationDialog
            label="Tem certeza que deseja excluir essa campanha?" 
            callback={async () => {
                await removeCampain(id)
                await fetchCampains()
                closeModal()
            }} 
        />)      
        openModal()
    }

    const handleEdit = () => {
        render(<CreateUpdateCampainForm campain={campains.data.find(c => c.id === id)}  />)      
        openModal()
    }

    return (
        <Popover >
            <PopoverTrigger>
                <EllipsisIcon className="text-slate-400" data-testid='action-menu' />
            </PopoverTrigger>
            <PopoverContent side="left" align="start" avoidCollisions className="p-0 max-w-[238px] ">
                <ActionItem label="Editar" icon={<Edit size={16}/>} onClick={() => handleEdit() }   />
                <ActionItem label="Excluir" icon={<Trash size={16} />} onClick={handleDelete}  />
            </PopoverContent>
        </Popover>
    )
}


