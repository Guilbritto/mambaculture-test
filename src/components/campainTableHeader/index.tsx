'use client'

import { CreateUpdateCampainForm } from "@/components/createUpdateCampainForm"
import { Button } from "@/components/ui/button"
import { useModal } from "@/context/useModal"

export const CampainTableHeader = () => {
    const {render, openModal} = useModal()
    
    const handleCreateCampain = () => {
        openModal()
        render(<CreateUpdateCampainForm />)
    }
    
    return (
        <div className="pt-2 px-2  border-b border-b-slate-200 pb-2 flex justify-between items-center mb-3 ">
            <h1 className="text-xl font-semibold text-primary"> Painel de controle </h1>
            <Button onClick={handleCreateCampain} className="rounded-lg bg-slate-700">Criar campanha</Button>
        </div >
    )
}