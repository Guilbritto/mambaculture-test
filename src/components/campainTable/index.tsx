'use client'

import { formatDate } from "@/lib/date"
import { Card } from "@/components/ui/card"
import {
    Table,
    TableBody,
    TableCell,
    TableHead,
    TableHeader,
    TableRow,
} from "@/components/ui/table"
import { Pagination, PaginationContent } from "@/components/ui/pagination"
import { Button } from "@/components/ui/button"
import {  ChevronLeft, ChevronRight } from "lucide-react"
import { Badge } from "@/components/ui/badge"
import { ActionMenu } from "@/components/actionMenu"
import { useCampain } from "@/context/useCampain"


export const CampainTable = () => {
    const {campains, fetchCampains} = useCampain()

    const handlePrevious = async () => {
        if(campains.prev === null) return
        await fetchCampains(campains.prev)
    }

    const handleNext = async () => {
        if(campains.next === null) return
        await fetchCampains(campains.next)
    }
    
    const isExpired = (endDate: string) => {
        const today = new Date()
        const end = new Date(endDate)
        return today > end
    }

    return (
        <Card>
            <Table className="">
                <TableHeader>
                    <TableRow>
                        <TableHead className="w-[100px]">Id</TableHead>
                        <TableHead>Nome</TableHead>
                        <TableHead className="w-[200px]">Data Inicio</TableHead>
                        <TableHead className="w-[200px]">Data Fim</TableHead>
                        <TableHead className="w-[200px]">Status</TableHead>
                        <TableHead className="w-[100px]">Ações</TableHead>
                    </TableRow>
                </TableHeader>
                <TableBody >
                    {campains.data?.map((campain) => (
                        <TableRow key={campain.id}>
                            <TableCell>#{campain.id}</TableCell>
                            <TableCell>{campain.name}</TableCell>
                            <TableCell>{formatDate(campain.startDate)}</TableCell>
                            <TableCell>{formatDate(campain.endDate)}</TableCell>
                            <TableCell>{
                                isExpired(campain.endDate) ? 
                                    <Badge className="bg-slate-700/60">Expirado</Badge> : 
                                    <Badge className="bg-slate-700" >Ativo</Badge>}
                            </TableCell>
                            <TableCell>
                                <ActionMenu id={campain?.id}/>
                            </TableCell>
                        </TableRow>
                    ))}
                </TableBody>
            </Table>
            <Pagination className="p-2 justify-end border-t">
                <PaginationContent >
                    <Button disabled={!campains.prev} className="bg-slate-700 hover:bg-slate-800" onClick={handlePrevious}>
                        <ChevronLeft />
                    </Button>
                    <Button disabled={!campains.next} className="bg-slate-700 hover:bg-slate-800" onClick={handleNext}>
                        <ChevronRight />
                    </Button>
                </PaginationContent>
            </Pagination>
        </Card>
    )
}