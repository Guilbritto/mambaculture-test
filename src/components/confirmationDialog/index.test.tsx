import React from 'react';
import { render } from '@testing-library/react';
import { ConfirmationDialog } from '@/components/confirmationDialog';
import { ModalProvider } from '@/context/useModal';

describe('ConfirmationDialog', () => {
    it('should match snapshot', () => {
        const { asFragment } = render(
            <ModalProvider>
                <ConfirmationDialog
                    label="Tem certeza que deseja excluir essa campanha?"
                    callback={jest.fn()}
                />
            </ModalProvider>

        );
        expect(asFragment()).toMatchSnapshot();
    });

    it('should call the callback', () => {
        const callback = jest.fn();
        const { getByText } = render(
            <ModalProvider>
                <ConfirmationDialog
                    label="Tem certeza que deseja excluir essa campanha?"
                    callback={callback}
                />
            </ModalProvider>
        );

        getByText('Continuar').click();
        expect(callback).toHaveBeenCalled();
    });

});