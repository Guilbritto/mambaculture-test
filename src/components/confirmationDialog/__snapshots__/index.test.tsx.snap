// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmationDialog should match snapshot 1`] = `
<DocumentFragment>
  <div>
    <h1
      class="text-lg font-semibold pb-3"
    >
      Atenção!
    </h1>
    <p
      class="text-muted-foreground text-sm font-normal"
    >
      Tem certeza que deseja excluir essa campanha?
    </p>
    <div
      class="flex justify-end gap-2 mt-5"
    >
      <button
        class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 border border-input bg-background shadow-sm hover:bg-accent hover:text-accent-foreground h-9 px-4 py-2"
      >
        Cancelar
      </button>
      <button
        class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground shadow hover:bg-primary/90 h-9 px-4 py-2"
      >
        Continuar
      </button>
    </div>
  </div>
</DocumentFragment>
`;
